File: app/src/test/java/com/sirloin/mtraceapiclient/internal/http/fixture/MockURL.java
package com.sirloin.mtraceapiclient.internal.http.fixture;

import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.net.URLStreamHandler;

@SuppressWarnings({"JavadocVariable", "MagicNumber"})
public final class MockURL {
    @SuppressWarnings({"MissingJavadocMethod"})
    public static final class MockURLStreamHandler extends URLStreamHandler {
        private final int mockResponseCode;

        public MockHttpURLConnection openConnection(final URL u) {
            MockURL.MockHttpURLConnection mockConn = new MockURL.MockHttpURLConnection(u);
            mockConn.setResponseCode(this.mockResponseCode);
            return mockConn;
        }

        public MockURLStreamHandler(final int mockResponseCode) {
            this.mockResponseCode = mockResponseCode;
        }
    }

    @SuppressWarnings({"MissingJavadocMethod"})
    public static final class MockHttpURLConnection extends HttpURLConnection {
        public void connect() {
        }

        public InputStream getInputStream() {
            return new MockURL.MockInputStream();
        }

        public InputStream getErrorStream() {
            return new MockURL.MockInputStream();
        }

        public void disconnect() {
        }

        public boolean usingProxy() {
            return false;
        }

        public void setResponseCode(final int mockResponseCode) {
            this.responseCode = mockResponseCode;
        }

        public MockHttpURLConnection(final URL u) {
            super(u);
        }
    }

    @SuppressWarnings({"MissingJavadocMethod"})
    public static final class MockInputStream extends InputStream {
        public int read() {
            return -1;
        }
    }
}

File: app/src/test/kotlin/com/sirloin/mtrace/util/http/mock/MockURL.kt
package com.sirloin.mtrace.util.http.mock

import java.io.InputStream
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLStreamHandler

class MockURL {
    class MockURLStreamHandler(private val mockResponseCode: Int) : URLStreamHandler() {
        override fun openConnection(u: URL?): MockHttpURLConnection {
            return MockHttpURLConnection(u).apply { responseCode = mockResponseCode }
        }
    }

    class MockHttpURLConnection(u: URL?) : HttpURLConnection(u) {

        // mock을 위한 구현체라 detekt의 빈 코드블록 경고는 무시하였습니다.
        @Suppress("EmptyFunctionBlock")
        override fun connect() {
        }

        override fun getInputStream(): InputStream {
            return MockInputStream()
        }

        override fun getErrorStream(): InputStream {
            return MockInputStream()
        }

        // mock을 위한 구현체라 detekt의 빈 코드블록 경고는 무시하였습니다.
        @Suppress("EmptyFunctionBlock")
        override fun disconnect() {
        }

        override fun usingProxy(): Boolean {
            return false
        }

        fun setResponseCode(mockResponseCode: Int) {
            responseCode = mockResponseCode
        }
    }

    class MockInputStream : InputStream() {
        override fun read(): Int {
            return -1
        }
    }

}
